### PR TITLE
[feat]: [CDS-47027] Asg NG Rolling Deploy Delegate Changes

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Asg NG Rolling Deploy Delegate Changes

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42080)

### CODE REVIEW

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingPrepareRollbackDataOutcome.java` line 30 at r1](https://reviewable.io/reviews/harness/harness-core/42080#-NKWxNdSDt6NYHSImNkL:-NKWxNdSDt6NYHSImNkM:bkn1x5r) ([raw file](https://github.com/harness/harness-core/blob/4ea2e8536bf92f977a6c8203d4c8d64d278d45fa/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingPrepareRollbackDataOutcome.java#L30)):*
> ```Java
> public class AsgRollingPrepareRollbackDataOutcome implements Outcome, ExecutionSweepingOutput {
>   String launchTemplateVersion;
>   String asgConfiguration;
> ```

This might not be extendible. We should keep complete data of each component of ASG.
Map<String, List<String>> could be a good candidate as we can store for all 4 (LT, ASGConfig, Scaling policies, scheduled update action) and can be extended in future.
For version checking, we can use the same response object and verify the version number from there.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackStep.java` line 151 at r1](https://reviewable.io/reviews/harness/harness-core/42080#-NKWxr5w0ZeTFxWGqE3X:-NKWxr5w0ZeTFxWGqE3Y:bx1twb2) ([raw file](https://github.com/harness/harness-core/blob/4ea2e8536bf92f977a6c8203d4c8d64d278d45fa/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackStep.java#L151)):*
> ```Java
>         AsgRollingRollbackRequest.builder()
>             .accountId(accountId)
>             .launchTemplateVersion(asgRollingPrepareRollbackDataOutcome.getLaunchTemplateVersion())
> ```

Here also, let's send the map and handler should be able to take care of what needs to be done.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java` line 412 at r1](https://reviewable.io/reviews/harness/harness-core/42080#-NKWy9bk7tilvEu_Ds48:-NKWy9bk7tilvEu_Ds49:b-ivt2qs) ([raw file](https://github.com/harness/harness-core/blob/4ea2e8536bf92f977a6c8203d4c8d64d278d45fa/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java#L412)):*
> ```Java
>           AsgRollingPrepareRollbackDataOutcome.builder();
> 
>       asgPrepareRollbackDataOutcomeBuilder.launchTemplateVersion(
> ```

same here. All places where we are individually sending these components can be updated to just use Map of configs.

**Comment by arvind-choudhary-h** [comment]
*[`960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java` line 20 at r1](https://reviewable.io/reviews/harness/harness-core/42080#-NKWza1nC7QomRxF44iM:-NKWza1oCHDKiQFmtWGY:b-9x6qff) ([raw file](https://github.com/harness/harness-core/blob/4ea2e8536bf92f977a6c8203d4c8d64d278d45fa/960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java#L20)):*
> ```Java
> public class AsgConfigurationRequest extends ManifestRequest {
>   public boolean useAlreadyRunningInstances;
>   Map<String, Object> overrideProperties;
> ```

can we move overrideProperties to `ManifestRequest`?

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java [hunk 1]
```diff
@@ -176,15 +217,165 @@ public CreateAutoScalingGroupResult createASG(
...
+    List<String> targetGroupARNs = createAutoScalingGroupRequest.getTargetGroupARNs();
+    if (isNotEmpty(targetGroupARNs)) {
+      targetGroupARNs.forEach(targetGroupARN -> {
+        AttachLoadBalancerTargetGroupsRequest attachLoadBalancerTargetGroupsRequest =
+            new AttachLoadBalancerTargetGroupsRequest().withAutoScalingGroupName(asgName).withTargetGroupARNs(
+                targetGroupARN);
+        asgCall(asgClient -> asgClient.attachLoadBalancerTargetGroups(attachLoadBalancerTargetGroupsRequest));
+      });
+    }
   }
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:379]: the method is too big, split in methods with tags, hooks, loadbalancers logic separate

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java [hunk 2]
```diff
@@ -278,6 +469,182 @@ public void waitReadyState(String asgName, Predicate<String> predicate, String o
     }
   }
 
+  public StartInstanceRefreshResult startInstanceRefresh(
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:472]: with small "s"

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java [hunk 3]
```diff
@@ -278,6 +469,182 @@ public void waitReadyState(String asgName, Predicate<String> predicate, String o
...
+
+    DescribeInstanceRefreshesResult describeInstanceRefreshesResult =
+        asgCall(asgClient -> asgClient.describeInstanceRefreshes(describeInstanceRefreshesRequest));
+    List<InstanceRefresh> instanceRefreshList = describeInstanceRefreshesResult.getInstanceRefreshes();
+
+    Set<String> statuses = instanceRefreshList.stream().map(InstanceRefresh::getStatus).collect(Collectors.toSet());
+    return statuses.size() == 1 && statuses.contains(INSTANCE_REFRESH_STATUS_SUCCESSFUL);
+  }
+
+  public void waitInstanceRefreshSteadyState(String asgName, String instanceRefreshId, String operationName) {
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:498]: no need to create this method, you can reuse `waitReadyState()` method. Just wrap your logic in a custom Predicate, like:
```
    String asgName = "ss";
    String arg2 = "arg2";
    Predicate<String> enhancedPredicate = (arg) -> checkInstanceRefreshReady(asgName, arg2);
    waitReadyState(asgName, enhancedPredicate, operationName);

```

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import io.harness.manifest.request.ManifestRequest;
+
+import java.util.List;
+import java.util.Map;
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+public class AsgConfigurationRequest extends ManifestRequest {
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java:18]: better to rename AsgConfigurationManifestRequest

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgInstanceRefreshRequest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+package io.harness.aws.asg.manifest.request;
+
+import io.harness.manifest.request.ManifestRequest;
+
+import java.util.List;
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+public class AsgInstanceRefreshRequest extends ManifestRequest {
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgInstanceRefreshRequest.java:17]: rename with Manifest

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgLaunchTemplateRequest.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+package io.harness.aws.asg.manifest.request;
+
+import io.harness.manifest.request.ManifestRequest;
+
+import java.util.List;
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+public class AsgLaunchTemplateRequest extends ManifestRequest {
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgLaunchTemplateRequest.java:17]: rename with Manifest

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgScalingPolicyRequest.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+package io.harness.aws.asg.manifest.request;
+
+import io.harness.manifest.request.ManifestRequest;
+
+import java.util.List;
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+public class AsgScalingPolicyRequest extends ManifestRequest {
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgScalingPolicyRequest.java:17]: rename with Manifest

File: 930-delegate-tasks/src/main/java/io/harness/delegate/aws/asg/AsgCanaryDeployCommandTaskHandler.java [hunk 8]
```diff
@@ -83,10 +84,8 @@ protected AsgCommandResponse executeTaskInternal(AsgCommandRequest asgCommandReq
       AutoScalingGroupContainer autoScalingGroupContainer =
           executeCanaryDeploy(asgSdkManager, asgStoreManifestsContent, serviceSuffix, nrOfInstances);
 
-      AsgCanaryDeployResult asgCanaryDeployResult = AsgCanaryDeployResult.builder()
-                                                        .region(asgInfraConfig.getRegion())
-                                                        .autoScalingGroupContainer(autoScalingGroupContainer)
-                                                        .build();
+      AsgCanaryDeployResult asgCanaryDeployResult =
+          AsgCanaryDeployResult.builder().autoScalingGroupContainer(autoScalingGroupContainer).build();
```
Comment (vitsafronovici) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/aws/asg/AsgCanaryDeployCommandTaskHandler.java:89]: why removed region?

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java [hunk 9]
```diff
@@ -144,7 +185,7 @@ public LaunchTemplate getLaunchTemplate(String asgName) {
         return null;
       }
       return resultList.get(0);
-    } catch (AmazonEC2Exception e) {
+    } catch (InvalidRequestException e) {
       // AmazonEC2Exception is thrown if LaunchTemplate is not found
```
Comment (vitsafronovici) [inline 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:189]: why changed that? how to return null?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java [hunk 10]
```diff
@@ -142,6 +146,15 @@ public StepResponse finalizeExecutionWithSecurityContext(Ambiance ambiance, Step
...
 
+    AsgRollingDeployOutcome asgRollingDeployOutcome =
+        AsgRollingDeployOutcome.builder()
+            .autoScalingGroupContainer(
+                asgRollingDeployResponse.getAsgRollingDeployResult().getAutoScalingGroupContainer())
+            .build();
+
+    executionSweepingOutputService.consume(ambiance, OutcomeExpressionConstants.ASG_ROLLING_DEPLOY_OUTCOME,
+        asgRollingDeployOutcome, StepOutcomeGroup.STEP.name());
+
```
Comment (vitsafronovici) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java:157]: I think we also need to save the manifests contents in sweeping output to be used for rollback

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java (hunk 9 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:189]: log.error(e.getMessage());
      Exception sanitizeException = ExceptionMessageSanitizer.sanitizeException(e);
      throw new InvalidRequestException(ExceptionUtils.getMessage(sanitizeException), sanitizeException);
    }
  }

this awscall method returns InvalidRequestException, so had to change to it.

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgScalingPolicyRequest.java (hunk 7 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgScalingPolicyRequest.java:17]: ok

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgLaunchTemplateRequest.java (hunk 6 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgLaunchTemplateRequest.java:17]: ok

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgInstanceRefreshRequest.java (hunk 5 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgInstanceRefreshRequest.java:17]: ok

File: 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java (hunk 4 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/manifest/request/AsgConfigurationRequest.java:18]: ok

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java (hunk 3 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:498]: ok

File: 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java (hunk 1 above)
Comment (lovish1999) [reply 960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java:379]: ok

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java (hunk 10 above)
Comment (lovish1999) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java:157]: hmm ok

**Comment by arvind-choudhary-h** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/aws/asg/AsgPrepareRollbackDataCommandTaskHandler.java` line 121 at r6](https://reviewable.io/reviews/harness/harness-core/42080#-NKrNHep2e4pEZUcc6_d:-NKrNHep2e4pEZUcc6_e:b-syopwv) ([raw file](https://github.com/harness/harness-core/blob/b6e4c4a3b53030f64d939f75f62033558b77b5df/930-delegate-tasks/src/main/java/io/harness/delegate/aws/asg/AsgPrepareRollbackDataCommandTaskHandler.java#L121)):*
> ```Java
>       String launchTemplateVersion = autoScalingGroup.getLaunchTemplate().getVersion();
>       String asgConfiguration =
>           asgMapper.createAutoScalingGroupRequestFromAutoScalingGroupConfiguration(autoScalingGroup);
> ```

should we use same Chaining logic we use to create. We should have GET methods.

**Review by vitsafronovici — APPROVED** [review]
:lgtm:

> Reviewed 10 of 32 files at r1, 6 of 18 files at r2, 1 of 3 files at r3, 11 of 13 files at r5, 1 of 1 files at r6, 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42080#-:-NKvtxTI8YJPLIMlooY8:bnfp4nl)* status: 39 file(s) to review. 1 of 2 approvals obtained.

**Comment by arvind-choudhary-h** [comment]
*[`960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java` line 623 at r7](https://reviewable.io/reviews/harness/harness-core/42080#-NKwBsrr8JbdR2ASJt7P:-NKwBsrr8JbdR2ASJt7Q:b-l6mwxj) ([raw file](https://github.com/harness/harness-core/blob/d194657936bd612e5fbb60116dde3cd6d50f0b7f/960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java#L623)):*
> ```Java
>   /*
>     public CreateLaunchTemplateRequest createLaunchTemplateRequestfromAsgName(String asgName) {
>       CreateLaunchTemplateRequest createLaunchTemplateRequest = new CreateLaunchTemplateRequest();
> ```

remove

**Comment by arvind-choudhary-h** [comment]
*[`960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java` line 613 at r7](https://reviewable.io/reviews/harness/harness-core/42080#-NKwBu7c0N-nAKXV8Yls:-NKwBu7c0N-nAKXV8Ylt:bs7re) ([raw file](https://github.com/harness/harness-core/blob/d194657936bd612e5fbb60116dde3cd6d50f0b7f/960-api-services/src/main/java/io/harness/aws/asg/AsgSdkManager.java#L613)):*
> ```Java
>         createAutoScalingGroupRequest.getNewInstancesProtectedFromScaleIn());
>     updateAutoScalingGroupRequest.setServiceLinkedRoleARN(createAutoScalingGroupRequest.getServiceLinkedRoleARN());
>     updateAutoScalingGroupRequest.setMaxInstanceLifetime(createAutoScalingGroupRequest.getMaxInstanceLifetime());
> ```

replace with mapper

**Comment by arvind-choudhary-h** [comment]
*[`960-api-services/src/main/java/io/harness/aws/asg/manifest/AsgManifestHandlerChainState.java` line 27 at r7](https://reviewable.io/reviews/harness/harness-core/42080#-NKwDyEBBGNgmHqnPZmW:-NKwDyEC0G-wUbq_SxoR:borvuqb) ([raw file](https://github.com/harness/harness-core/blob/d194657936bd612e5fbb60116dde3cd6d50f0b7f/960-api-services/src/main/java/io/harness/aws/asg/manifest/AsgManifestHandlerChainState.java#L27)):*
> ```Java
>   private String launchTemplateVersion;
>   private AutoScalingGroup autoScalingGroup;
>   private Map<String, List<String>> prepareRollbackDataAsgStoreManifestsContent;
> ```

can rename

**Comment by arvind-choudhary-h** [comment]
*[`960-api-services/src/main/java/io/harness/aws/asg/AsgMapper.java` line 36 at r7](https://reviewable.io/reviews/harness/harness-core/42080#-NKwEvS47m8zcKpsFsZU:-NKwEvS47m8zcKpsFsZV:b-jswbgb) ([raw file](https://github.com/harness/harness-core/blob/d194657936bd612e5fbb60116dde3cd6d50f0b7f/960-api-services/src/main/java/io/harness/aws/asg/AsgMapper.java#L36)):*
> ```Java
>       throws JsonProcessingException {
>     CreateAutoScalingGroupRequest createAutoScalingGroupRequest = new CreateAutoScalingGroupRequest();
>     createAutoScalingGroupRequest.setAutoScalingGroupName(autoScalingGroup.getAutoScalingGroupName());
> ```

need to update.

**Review by arvind-choudhary-h — COMMENTED** [review]
> Reviewed 13 of 32 files at r1, 7 of 18 files at r2, 1 of 3 files at r3, 11 of 13 files at r5, 1 of 1 files at r6, 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42080)* status: 1 of 2 approvals obtained.

**Review by vitsafronovici — APPROVED** [review]
> Reviewed 8 of 12 files at r8, 4 of 5 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42080)* status: 13 file(s) to review. 1 of 2 approvals obtained.

**Review by arvind-choudhary-h — APPROVED** [review]
> Reviewed 9 of 12 files at r8, 4 of 5 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42080)* status: :shipit: complete! ✓

**Pull request merged**